services/queued/internal/queue/queue.go

```diff
@@ -228,249 +234,271 @@ func zipOutputToBase64(output_path string) (string, error) {
 
+	var lastGoodRead = time.Now()
+
+	args := GetArguments()
+
 	for {
-		msg, more := <-queue
-		if !more {
-			log.Println("<6>Message queue empty and closed.")
+
+		select {
+
+		case <-ctx.Done(): // if cancel() execute
+			log.Println("<6>Gracefully canceling ProcessMessages loop...")
 			return
-		}
 
-		log.Println("<6>Start processing of message.")
+		default:
 
-		func() {
-			// This is to always log last request, in case something went terribly bad
-			defer func() {
-				r := recover()
-				if r != nil {
-					log.Panicf("<3>Panicing on message %v", msg.MetaData) // just fail quickly
-				} else {
-					log.Println("<6>Finished processing of message.")
-				}
-			}()
+			log.Println("<6>Get message via http request")
 
-			var outputVolume docker.DockerVolume
-			// we do not want to delte the tmp files, it may be useful,
-			// also the line below will not do the trick, since args
-			// are bound when defer is called, not when the function is
-			// called, so it will call Cleanup on empty path :(
-			//defer outputVolume.Cleanup()
-
-			result := func() string {
-				ctx := context.Background()
-
-				config := docker.DockerConfig{
-					Volumes:     []docker.DockerVolume{},
-					DockerImage: msg.DockerImage,
-					Timeout:     uint16(msg.MaxRunTime),
-					Memory:      int64(msg.Memory * 1024 * 1024),
-					Username:    arguments.DockerUsername,
-					Password:    arguments.DockerPassword,
-				}
+			resp, err := http.Get(args.MqReadServiceAddr)
+			if err != nil {
+				log.Println(err)
+				continue
+			}
 
-				docker.PullImage(ctx, config)
+			msg := requestMessage{
+				MaxRunTime: 500, // 300 for test and 200 for build
+				Memory:     2048,
+			}
 
-				env, err := docker.GetDockerEnv(ctx, msg.DockerImage)
-				if err != nil {
-					return err.Error()
-				}
+			readOk := func() (isOk bool) {
+				defer resp.Body.Close()
 
-				outputVolume = getVolume(strings.TrimSpace(env["OUTPUT_PATH"]), false)
+				log.Println("<6>Status code returned from HTTP GET ", resp.StatusCode)
 
-				submissionVolume := getVolume(strings.TrimSpace(env["SUBMISSION_PATH"]), true)
-				defer submissionVolume.Cleanup()
+				if resp.StatusCode == 404 {
+					log.Println("<6>Empty response. No messages in input queue.")
 
-				for file, content := range msg.Files {
-					path := filepath.Join(submissionVolume.HostPath, file)
+					// check for idle time
+					diff := time.Since(lastGoodRead).Seconds()
+					log.Println("<6>Idle time: ", diff, " s.")
 
-					err := ioutil.WriteFile(path, []byte(content), 0644)
-					if err != nil {
-						log.Println("<3>Cannot write submission file", err)
-						return err.Error()
+					if diff > arguments.MaxIdleTime {
+
+						log.Println("<6>max idle time (", args.MaxIdleTime, ") was exceeded.")
+						log.Println("<6>We proceed to shutdown...")
+
+						exit()
 					}
+
+					return false
+				}
+
+				if (resp.StatusCode < 200 || resp.StatusCode >= 300) && resp.StatusCode != 404 {
```

Comment:
I guess no need to check for 404 here, since we deal with it above.

Reply:
Fixed